nav2_ceres_costaware_smoother/include/nav2_ceres_costaware_smoother/options.hpp

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2021 RoboTech Vision
+// Copyright (c) 2020, Samsung Research America
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License. Reserved.
+
+#ifndef NAV2_CERES_COSTAWARE_SMOOTHER__OPTIONS_HPP_
+#define NAV2_CERES_COSTAWARE_SMOOTHER__OPTIONS_HPP_
+
+#include <string>
+#include <vector>
+#include "rclcpp_lifecycle/lifecycle_node.hpp"
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_ceres_costaware_smoother
+{
+
+/**
+ * @struct nav2_smac_planner::SmootherParams
+ * @brief Parameters for the smoother cost function
+ */
+struct SmootherParams
+{
+  /**
+   * @brief A constructor for nav2_smac_planner::SmootherParams
+   */
+  SmootherParams()
+  {
+  }
+
+  /**
+   * @brief Get params from ROS parameter
+   * @param node_ Ptr to node
+   * @param name Name of plugin
+   */
+  void get(rclcpp_lifecycle::LifecycleNode * node, const std::string & name)
+  {
+    std::string local_name = name + std::string(".");
+
+    // Smoother params
+    double minimum_turning_radius;
+    nav2_util::declare_parameter_if_not_declared(
+      node, name + ".minimum_turning_radius", rclcpp::ParameterValue(0.4));
+    node->get_parameter(name + ".minimum_turning_radius", minimum_turning_radius);
+    max_curvature = 1.0f / minimum_turning_radius;
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "w_curve", rclcpp::ParameterValue(30.0));
+    node->get_parameter(local_name + "w_curve", curvature_weight);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "w_cost", rclcpp::ParameterValue(0.015));
+    node->get_parameter(local_name + "w_cost", costmap_weight);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "w_cost_cusp", rclcpp::ParameterValue(0.04));
+    node->get_parameter(local_name + "w_cost_cusp", cusp_costmap_weight);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "cusp_zone_length", rclcpp::ParameterValue(-1.0));
```

Comment:
Ditto

Reply:
It has been empirically found that the cost around cusp needs to increase gradually (linearly) towards the cusp. `cusp_zone_length` parameter defines the area (and thus the slope as well) of increase

Reply:
As you guessed, -1 means "disabled", i.e. all points use `w_cost` only. My decision was to disable it so that during tuning users first focus on `w_cost` and enable the cusp handling afterwards, choosing a reasonable `w_cost_cusp` value, relatively to their tuned `w_cost`. The cusp cost is enabled in the `README.md` example, I just wanted to spare users who tune from scratch (and don't read whole docs) from confusions like "I turned `w_cost` totally off, why the path is still being pushed away from obstacles at some places?" (note: it can also happen that `w_cost_cusp` remains lower than `w_cost` if not adjusted properly, again, causing confusions).

Just telling my thread of thoughts, I have no problem to change the default if you decide.

Reply:
Just came to mind, that a `w_cost_cusp_multiplier` could be used instead of the absolute parameter. This would prevent confusions and thus the cusp cost could be enabled by default with no problems.

Reply:
I like that idea - a multiple on the cost parameter to apply at cusps. 

I think it should also be defaulted to "on" as long as the defaults you provide in the source code are "reasonable" for out of the box users. If folks don't need to change the parameters (or change them very much) its probably not going to be come an issue. If users want to tune from absolute scratch, then this isn't the only package they're going to get wacky results in if they don't understand what the dials do.